velox/exec/NestedLoopJoinBuild.cpp

```diff
@@ -105,6 +105,10 @@ void NestedLoopJoinBuild::noMoreInput() {
     }
   }
 
+  while (dataVectors_.size() > 1) {
+    dataVectors_[0]->append(dataVectors_[dataVectors_.size()-1].get());
```

Comment:
No need to merge the vector, you just need to change the loop order in probe to do the probe side first

Reply:
@karteekmurthys : Agree with Jimmy. We need to change the probe loop to generate the indices differently.

Reply:
One approach I tried was generating all the probe side indices for total build vector size (sum of all build vectors at the time of probing). Since we compute `buildSize` as sum of all build vectors, this also takes care of generating the build side indices. Now, all we need to handle is projections which expects a single RowVector for probe and build side. It seems like at this point we are forced to merge all our build vectors to be projected at once. That is why I chose to merge vectors in NestedLoopBuild side. Please CMIW there could be a better way to map indices to vectors.

```
  projectChildren(
      projectedChildren,
      buildVectors_.value()[buildIndex_], <- Single RowVector, but we need to project multiple into single output.
      buildProjections,
      numOutputRows,
      buildIndices_);
```

Reply:
Synced with @Yuhta will try another approach and update this PR.

Reply:
Discussed offline, merging build vectors help avoiding small output batches so might worth doing here.  But `BaseVector::append` is not the right implementation here as it's too slow.  Can you create a static method in `BaseVector` like this?

```c++
VectorPtr BaseVector::merge(const std::vector<VectorPtr>& vectors, memory::MemoryPool* pool) {
  VELOX_CHECK(!vector.empty());
  auto& type = vectors[0].type();
  vector_size_t size = 0;
  for (auto& vector : vectors) {
    VELOX_CHECK_EQ(*vector->type(), *type);
    size += vector->size();
  }
  auto result = create(type, size, pool);
  size = 0;
  for (auto& vector : vectors) {
    result->copy(vector.get(), size, 0, vector->size());
    size += vector->size();
  }  
  return result;
}
```

Reply:
I think about it again, merging will produce batches larger than we currently use, so still have a risk of memory-wise.  We probably should still stick to changing looping order approach.

Reply:
@Yuhta, @mbasmanova : There is another option here. This grouping behavior from NestedLoopJoinNode is needed because of the use of StreamingAggregation in the original query.

We could change the LocalPlanner to use HashAggregation instead of the StreamingAggregation when used over a NestedLoopJoin instead for the correctness.

To elaborate, the original issue came from incorrect results in the following query:

`SELECT count(*) FROM orders o WHERE EXISTS(SELECT 1 FROM orders i WHERE o.orderkey < i.orderkey AND i.orderkey % 1000 = 0);`

count(*) needed aggregation and the subquery needed NestedLoopJoin in Prestissimo.

In Presto Java -- the same JoinNode translates to LookupJoin which imposes a sorting on the HashBuild side in this code https://github.com/prestodb/presto/blob/master/presto-main/src/main/java/com/facebook/presto/sql/planner/LocalExecutionPlanner.java#L2436. So it justified the usage of StreamingAggregation.

We don't have an equivalent for such a LookupJoin in Velox and used NestedLoopJoin.

Since StreamingAggregation above NestedLoopJoin requires pre-grouping, it didn't give correct results.

As demonstrated in @karteekmurthys example

```
In NestedLoop Probe side if the input is

1, 1
2, 2

and build side you have 2 vectors

Vector 1 : 1
Vector 2: 2

The output is going to be

1, 1, 1
2, 2, 1
1, 1, 2
2, 2, 2

which is ungrouped.
```
This double counts the groups for both 1 and 2 keys.

We could change in local planning to use HashAggregation instead of StreamingAggregation above NestedLoopJoin to avoid the mis-counting.